Tools/scripts/freeze_modules.py

```diff
@@ -293,38 +450,65 @@ def replace_block(lines, start_marker, end_marker, replacements, file):
     return lines[:start_pos + 1] + replacements + lines[end_pos:]
 
 
-def regen_frozen(specs, dest=MODULES_DIR):
-    if isinstance(dest, str):
-        frozen, frozenids = resolve_frozen_files(specs, destdir)
-    else:
-        frozenids, frozen = dest
+def regen_manifest(modules):
+    header = 'module ispkg source frozen checksum'.split()
+    widths = [5] * len(header)
+    rows = []
+    for mod in modules:
+        info = mod.summarize()
+        row = []
+        for i, col in enumerate(header):
+            value = info[col]
+            if col == 'checksum':
+                value = value[:12]
+            elif col == 'ispkg':
+                value = 'YES' if value else 'no'
+            widths[i] = max(widths[i], len(value))
+            row.append(value or '-')
+        rows.append(row)
+
+    modlines = [
+        '# This file is auto-generated by Tools/scripts/freeze_modules.py.',
```

Comment:
Per our discussion, add an explanation of what it's used for.

Honestly I'm not yet convinced that it's actually useful. I worry it will end up just being yet another chore when you edit a .py file that happens to be frozen.

Regardless, does the MANIFEST file need to be added to the generated files in .gitattributes?

Reply:
> Per our discussion, add an explanation of what it's used for.

Done!

> Honestly I'm not yet convinced that it's actually useful. I worry it will end up just being yet another chore when you edit a .py file that happens to be frozen.

Yeah, it might not be worth using it for that (and for the moment it isn't, since I've added the frozen .h back).  Personally, I find the manifest to be useful on its own as a clean list of all the frozen modules, but don't see it as necessary.  So if you think the file doesn't add enough value, I don't mind dropping it.

> Regardless, does the MANIFEST file need to be added to the generated files in .gitattributes?

Good point!  I've done that.